File: projects/composablekernel/python/ck_dsl_c/tests/parity/moe_sorting_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/moe_sorting_emit.py -- Python reference emitter for the
# MoE-sorting parity harness. Selects one of the sampled configs by argv[2]
# (the config index) and the phase by argv[1] ("hist"/"scan"/"scatter"),
# builds the MoeSortingSpec, builds the kernel via the matching
# build_moe_sort_* and prints lower_kernel_to_llvm(arch='gfx950') to stdout so
# it can be byte-compared with the C emitter moe_sorting_emit.c.
import sys

from ck_dsl.instances.common.moe_sorting import (
    MoeSortingSpec,
    build_moe_sort_histogram,
    build_moe_sort_scan,
    build_moe_sort_scatter,
)
from ck_dsl import lower_kernel_to_llvm


def _spec(idx: int) -> MoeSortingSpec:
    if idx == 0:
        return MoeSortingSpec(tokens=2, topk=8, experts=8, block_size=64)
    if idx == 1:
        return MoeSortingSpec(tokens=16, topk=4, experts=32, block_size=256)
    if idx == 2:
        return MoeSortingSpec(tokens=32, topk=8, experts=64, block_size=256)
    if idx == 3:
        return MoeSortingSpec(tokens=128, topk=2, experts=32, block_size=512)
    if idx == 4:
        return MoeSortingSpec(tokens=8, topk=16, experts=16, block_size=128)
    if idx == 5:
        return MoeSortingSpec(tokens=2, topk=8, experts=64, block_size=256)
    raise SystemExit(f"unknown config index {idx}")


_BUILD = {
    "hist": build_moe_sort_histogram,
    "scan": build_moe_sort_scan,
    "scatter": build_moe_sort_scatter,
}


def main() -> int:
    if len(sys.argv) < 3:
        sys.stderr.write(
            "usage: moe_sorting_emit.py <phase hist|scan|scatter> <config_index 0..5>\n"
        )
        return 2
    phase = sys.argv[1]
    idx = int(sys.argv[2])
    if phase not in _BUILD:
        raise SystemExit(f"unknown phase {phase}")
    spec = _spec(idx)
    kernel = _BUILD[phase](spec, arch="gfx950")
    text = lower_kernel_to_llvm(kernel, arch="gfx950")
    sys.stdout.write(text)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: projects/composablekernel/python/ck_dsl_c/tests/parity/moe_sorting_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/moe_sorting_emit.c -- C-side emitter for the MoE-sorting
 * instance parity harness. Selects one of the sampled configs by argv[2] (the
 * config index) and the phase by argv[1] ("hist"/"scan"/"scatter"), builds
 * ckc_moe_sorting_spec_t identically to the Python emitter moe_sorting_emit.py,
 * validates via ckc_moe_sorting_is_valid_spec, builds into a fresh IRBuilder via
 * the matching ckc_build_moe_sort_*_new C build entry, lowers via
 * ckc_lower_kernel_to_llvm (arch gfx950, flavor AUTO) and prints the .ll to
 * stdout so the two outputs can be byte-compared.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ckc/ir.h"
#include "ckc/lower_llvm.h"
#include "ckc/instance_moe_sorting.h"

/* Fill `spec` for config index `idx`. Returns 0 on success, -1 if unknown. */
static int make_spec(int idx, ckc_moe_sorting_spec_t *spec) {
    *spec = ckc_moe_sorting_spec_default();

    switch (idx) {
    case 0:
        spec->tokens = 2;   spec->topk = 8;  spec->experts = 8;
        spec->block_size = 64;
        break;
    case 1:
        spec->tokens = 16;  spec->topk = 4;  spec->experts = 32;
        spec->block_size = 256;
        break;
    case 2:
        spec->tokens = 32;  spec->topk = 8;  spec->experts = 64;
        spec->block_size = 256;
        break;
    case 3:
        spec->tokens = 128; spec->topk = 2;  spec->experts = 32;
        spec->block_size = 512;
        break;
    case 4:
        spec->tokens = 8;   spec->topk = 16; spec->experts = 16;
        spec->block_size = 128;
        break;
    case 5:
        spec->tokens = 2;   spec->topk = 8;  spec->experts = 64;
        spec->block_size = 256;
        break;
    default:
        return -1;
    }
    return 0;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        fprintf(stderr,
                "usage: %s <phase hist|scan|scatter> <config_index 0..5>\n",
                argv[0]);
        return 2;
    }
    const char *phase = argv[1];
    int idx = atoi(argv[2]);

    ckc_moe_sorting_spec_t spec;
    if (make_spec(idx, &spec) != 0) {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    const char *arch = "gfx950";

    /* Validate the spec (mirrors is_valid_spec). */
    char reason[CKC_ERR_MSG_CAP];
    reason[0] = 0;
    if (!ckc_moe_sorting_is_valid_spec(&spec, arch, reason, sizeof reason)) {
        fprintf(stderr, "invalid spec: %s\n", reason);
        return 1;
    }

    /* Select the phase build entry. */
    ckc_kernel_def_t *(*build_new)(ckc_ir_builder_t *,
                                   const ckc_moe_sorting_spec_t *,
                                   const char *) = NULL;
    if (strcmp(phase, "hist") == 0) {
        build_new = ckc_build_moe_sort_histogram_new;
    } else if (strcmp(phase, "scan") == 0) {
        build_new = ckc_build_moe_sort_scan_new;
    } else if (strcmp(phase, "scatter") == 0) {
        build_new = ckc_build_moe_sort_scatter_new;
    } else {
        fprintf(stderr, "unknown phase '%s'\n", phase);
        return 2;
    }

    /* Init IRBuilder with spec.kernel_name(<phase>) and build into it. */
    ckc_ir_builder_t b;
    ckc_kernel_def_t *kernel = build_new(&b, &spec, arch);
    if (!kernel) {
        fprintf(stderr, "build failed: %s\n", b.err);
        ckc_ir_builder_free(&b);
        return 1;
    }

    /* lower_kernel_to_llvm(kernel, arch='gfx950'). */
    char *llvm_text = NULL;
    ckc_status_t st = ckc_lower_kernel_to_llvm(
        kernel, CKC_LLVM_FLAVOR_AUTO, arch, &llvm_text);
    if (st != CKC_OK || !llvm_text) {
        fprintf(stderr, "lower failed: status=%d\n", (int)st);
        ckc_ir_builder_free(&b);
        return 1;
    }
    fputs(llvm_text, stdout);
    free(llvm_text);
    ckc_ir_builder_free(&b);
    return 0;
}
